# Fundamental Algorithms
(Five most commonly used **supervised learning algorithms** will be explained here). These are :-

**1. Linear Regression   
2. Logistic Regression   
3. Decision Tree Learning  
4. Support Vector Machine  
5. k Nearest Neighbors**  

---



# 1. Linear Regression 

A learning algorithm that learns a model which is a linear combination of features of the input examples and uses the model to output real-valued target when an input in given.

- Commanly used for regression task
- Simple and effective

## Problem Statement


- **Given** : *Dataset* of $N$ labelled examples - $\{x_i,y_i\}_{i=1}^N$ 

    **$x_i$** - D-dimensional feature vector , $y_i$ - real-valued target

- **Goal** : Build a model ($f_{\boldsymbol{w},b}$) given by,

    $f_{\boldsymbol{w},b}(x)= \boldsymbol{w^*x} + b^*$

    where $w^*$ and $b^*$  the optimal values that makes the most accurate predictions


    **Predict**: Unknown $y$ using : **$y=f_{\boldsymbol{w},b}(x_{new})$** ; 

- **Note**:-
    - It predicts real-values not classes.
    - It tries to decrease the error.



**Visualization**  

<img src="Images/Linear_regression_1D.png" width="500">

Regression model for :-
- $1D$ - a line
- $2D$ - a plane
- $D$ - dimensions - a hyperplane of $D$-dimensions

**Note:** It is different from SVM (which has hyperplane of $D - 1$ dimensions).

---



## Solution

- **Objective function** - function that we minimize or maximize to find the optimal values of the original function (the model).

- We want to find the optimal values of $w$ and $b$ by minimising the prediction error. It is given by the **cost function** (or empirical risk).

    $$
    \frac{1}{N}\sum_{i=1}^N(f{w,b}(x_i)-y_1)^2
    $$

    where $(f{w,b}(x_i)-y_1)^2$ is called the loss function (or squared error loss)

- **Loss function** is a measure of how wrong the predicted value is from the real value (i.e penalty for misclassification).



- **Why Linear** ?
    - It's simple.
    - Rarely **overfits**.

- **Why use squared loss? Why not absolute or cubed?**

    - Absoulte error? Cubed error? All are valid. But we use squared because 
        - It exaggerates larger errors - so penalize big mistakes more

        - Has smooth , continuous derivative - Optimization becomes easier. Simpler to compute closed form solution using linear algebra.

---



- **Gradient descent** : a complex numerical optimization method (discussed later) that can help in finding the optimal values of the parameters.

- **Overfitting** (high accuracy on training data, poor performance on unseen data)

<img src="Images/Linear_regression_1D.png" width="550">
<img src="Images/Overfitting.png" width="500"> 


---

## Optimization 

- Optimization: Minimizing cost of a function
- can be done by setting the gradient of the cost function to zero or other methods like Gradient descent.
- This gives optimal values - $w$* , $b$* in case of Linear Regression.

    $$
    C(w,b) = \frac{1}{N}\sum_{i=1}^N(f_{w,b}(x_i)-y_1)^2  
    $$

    $$
    \frac{\partial{C}}{\partial{w}}= 0 ; \frac{\partial{C}}{\partial{b}}= 0
    $$

---



# 2. Logistic Regression

- Not really a regression, but classification. 
- Named so - because its mathematical form is similar to linear regression.

## Problem Statement 

For binary classification (can also be extended to multiclass)

- **Given** : *Dataset* of $N$ labelled examples - $\{x_i,y_i\}_{i=1}^N$ where $y_i\in\{0,1\}$ 

- **Goal** : Build a linear model that predicts whether an input belongs to class 0 or 1.

    - Problem: Can't use linear expression like **$wx_i + b$** as it spans from -$\infty$ to +$\infty$ while $y_i$ only has two values.

    - Solution: Use a simple continuous function whose codomain is (0,1)
    
---



## Logistic Regression Model

Defined as, 

$$

f_{w,b}(x)= \frac{1}{1 + e^{-(wx + b)}}

$$

- a logistic function that give values between 0 and 1 
- can be interpreted as $Pr(y=1\mid x)$
- if $f(x) >= 0.5$, predict class 1
- if $f(x) < 0.5$, predict class 0

---



## Solution

(Again we want to find the optimal values $w^*$ & $b^*$)

- Unlike linear regression ( which minimizes **MSE**), 
 Logistic regression uses **maximum likelihood** estimation (which maximize the likelihood of the training data according to our model). Given by, 

 $$
 L_{w,b}= \prod_{i=1}^N f_{w,b}(x_i)^{y_i} (1-f_{w,b}(x_i))^{1-y_i} 
 $$
 $$
 f_{w,b}(x)\;when\;y_i=1 , (1-f_{w,b}(x))\;otherwise
 $$

- In practice, we take *log-likelihood* instead because of :-
    - presence of $exp$ function, taking $log$ makes it easier to compute
    - turns products into sum
    - leads to same optimization

$$
\log L_{w,b}= \sum_{i=1}^N y_i\;lnf_{w,b}(x) + (1-y_i)\;ln(1-f_{w,b}(x))
$$

## Optimization

- Unlike linear regression, it doesn't have a closed forms solution

- **Gradient descent** is used to optimise such problems

---



# 3. Decision Tree Learning

- A learning algorithm that makes a **non-parametric model** given by a decision tree to predict which class the input feature vector belongs to.
- **Decision Tree**: 
    - an acyclic graph
    - where at each node of the graph a specific feature $j$ of the feature vector is examined.
    - if the value of the feature is below threshold, then left branch is followed. Else, the right branch is followed. 
    - Leaf node represents class predictions.

<img src="Images/Decision_Tree_3.png" width = 500>



## Problem statement

**Given**: *Dataset* of $N$ labelled examples - $\{x_i,y_i\}_{i=1}^N$ where $y_i\in\{0,1\}$ 

**Goal**: To build a decision tree (model) that can predict the label of the the class to which the input feature vector belongs to.

- The tree is built from the data
- Initially, all the training data is at the root node.
- Repeatedly split the data by choosing the best feature & threshold.




## Model function

- Initially, start with a constant model $f_{ID3}^S$ given by,
$$
f_{ID3}^S = \frac{1}{\mid S \mid} \sum_{(x,y)\in S}y
$$

- Predicts same value for any input $x$.




### Splitting the Data

- Try all features $j = 1, \dots, D$ and all thresholds $t$.
- Split into:
  - $S^- = \{(x, y) \mid x^{(j)} < t\}$
  - $S^+ = \{(x, y) \mid x^{(j)} \geq t\}$
- Choose split that gives **lowest entropy**.

#### Entropy
- measure of uncertainty about a random variable. Given by,
$$
H(S) = -f_{ID3} \ln f_{ID3} - (1 - f_{ID3}) \ln(1 - f_{ID3})
$$

- Minimum Entropy = 0 ; all labels are same (no uncertainty).
- Maximum Entropy = 1 ; labels are split 50-50 (maximum uncertainty).

- Entropy after a Split :
    $$
    H(S^-, S^+) = \frac{|S^-|}{|S|} H(S^-) + \frac{|S^+|}{|S|} H(S^+)
    $$

    - Weighted average of entropies of both subsets.
    - Best split minimizes this value.


### Stop Splitting

Tree growth stops at a leaf node, if :

- All labels in node are same.
- No good split is found.
- Entropy reduction < $\varepsilon$ (found experimentally).
- Tree reaches maximum depth $d$ (found experimentally).



## Optimization Criterion

- The algorithm implicitly connects to optimizing a  **log-likelihood** :
$$
\frac{1}{N} \sum_{i=1}^N y_i \ln f_{ID3}(x_i) + (1 - y_i) \ln(1 - f_{ID3}(x_i))
$$

 ## Limitations of ID3

- Makes local decisions, doesn’t guarantee globally optimal tree.
- Can be improved using Backtracking (but can possibly increase the time taken to build the model) .

---



# 4. Support Vector Machines (SVMs)

- Finds the best hyperplane that separates classes with the widest margin. 

- We’ve already seen the basic idea. Now we explore how SVM handles :  
  1. **Noisy data** (imperfect separation)  
  2. **Non-linear boundaries** (inherently complex patterns)  

---



## Hard-Margin SVM
(already explained before)

- Constraints for perfectly separable data:  
  $$
  \begin{cases}
  w x_i - b \ge +1 & \text{if } y_i = +1 \\[6pt]
  w x_i - b \le -1 & \text{if } y_i = -1
  \end{cases}
  $$

- **Objective:** maximise margin  → minimise $\tfrac12 \|w\|^2$.  
  $$
  \min_{w,b}\; \tfrac12 \|w\|^2
  \quad\text{s.t.}\quad
  y_i\bigl(w x_i - b\bigr) \ge 1\;\; \forall i
  $$

---



## 1. Dealing with Noise 

(Linearly *Non-Separable* by Outliers)

- Some points cannot be separated perfectly due to noise or mislabeling.  
- Hard-margin SVM fails to find a perfect hyperplane.

---



### Soft-Margin & Hinge Loss

- **Hinge loss:**  
  $$
  \max\!\bigl(0,\; 1 - y_i(w^\top x_i - b)\bigr)
  $$

  It is zero when the constraints are satisfied (i.e $wx_i$ lies on the correct side)

- **New cost function** :  
  $$
  \min_{w,b}\; C\|w\|^2 + \frac1N \sum_{i=1}^{N}
           \max\!\bigl(0,\; 1 - y_i(w x_i - b)\bigr)
  $$

- **Hyper-parameter \(C\):**

    Chosen experimentally
    - High \(C\) → penalize errors, smaller margin  
  - Low \(C\) → allow some errors, larger margin → better generalisation

---



## 2. Dealing with Inherent Non-Linearity

- Even noise-free data may need a **curved** boundary 
- cannot be seperated by a hyperplane in the original space.

### Solution : Map to Higher Dimension

- Use a mapping $\phi : x \to \phi(x)$ where $\phi(x)$ is a vector in higher dimension so that data becomes linearly separable. 

- Example : for 2D point $(q,p)$:  
  $$
  \Phi(q,p) = \bigl(q^2,\; \sqrt2\,qp,\; p^2\bigr)
  $$



<img src="Images/SVM(3D).png" width= 525> 

<img src="Images/SVM (2D).png" width= 500>


- In new space, a plane can separate the classes.



## SVM Optimization

- Optimization uses **Lagrange multipliers**  
- Dual form (simplified):

$$
\max_{\alpha} \sum_{i=1}^{N} \alpha_i - \frac{1}{2} \sum_{i=1,j=1}^N \alpha_i \alpha_j y_i y_j (x_i.x_j)
$$

Subject to:
$$
\sum_{i=1}^{N} \alpha_i y_i = 0, \quad \alpha_i \geq 0
$$




## Kernel Trick

- For data in higher dimensional form, we nedd to replace $x.x'$ by $\Phi(x_i) \Phi(x_k)$. It would it costly to do.
- To avoid computation of $\Phi(x)$ explicitly.  
    - Replace dot-product $\Phi(x_i) \Phi(x_k)$ by kernel $k(x_i,x_k)$ where $k(x_i,x_k)$ gives the same result as $\Phi(x_i) \Phi(x_k)$ 

- **Common kernels**: 

  - Polynomial: $k(x,x') = (x.x')^n$  
  - RBF: $k(x,x') = \exp \left(\frac{-\|x-x'\|^2} {2\sigma^2}\right)$

    where $\|x-x'\|^2$ is the squared **Euclidean distance** between two feature vectors.



## Final Decision Function

$$
f(x) = \operatorname{sign}\Big(\sum_{i \in SV} \alpha_i y_i k(x_i, x) - b\Big)
$$

- Only **support vectors** (with $\alpha_i > 0$) influence the decision.

---



# 5. k-Nearest Neighbors (kNN)

- **Non-parametric** and **instance-based** algorithm  
- Does **not discard** training data instead uses it as a model  
- For a new input $x$:  
  - Finds $k$ closest training examples  
  - **Returns:**  
    - Majority label (classification)  
    - Average label (regression)



## How Distance Is Measured

- Requires a ***distance metric***
- Common choices:
  - **Euclidean distance**
  - **Negative cosine similarity**
  - **Chebyshev distance**
  - **Mahalanobis distance**
  - **Hamming distance**
- *Distance metric* and $k$ are **hyperparameters**




## Cosine Similarity

$$
\text{s}(x_i, x_k) = \cos(\theta) = 
\frac{\sum_{j=1}^{D} x_i^{(j)} x_k^{(j)}}{
\sqrt{\sum_{j=1}^{D} (x_i^{(j)})^2} \cdot
\sqrt{\sum_{j=1}^{D} (x_k^{(j)})^2}
}
$$

- 1 → same direction  
- 0 → orthogonal  
- –1 → opposite direction  
- Multiply by –1 to use as **distance**

---


## Cost function?
- KNN was introduced as a lazy learning algorithm. There was no cost function defined or minimized during training — because there is no training.
- An implicit cost function **Given by Li & Yang, 2003**

### Prediction as Local Linear Classifier

Assume:
- Binary classification: $y \in \{0, 1\}$  
- Normalized vectors

$$
w_x = \sum_{(x', y') \in R_k(x)} y' x'
$$
 where $R_k(x)$ is the set of k nearest neighbors to the input example $x$
 
- Sum of feature vectors of **positive-labeled** neighbors  
- Decision based on cosine similarity:
$$
\text{predict} = \text{sign}(x w_x)
$$


### Cost Function
**Given by Li & Yang, 2003**

$$
L = -\sum_{(x', y') \in R_k(x)} y' x' w_x + \frac{1}{2} \|w_x\|^2
$$

- Optimizing this gives same $w_x$ as before  
- kNN approximates a **local linear model**



*(Using all these five learning models for predicting)*


# Python Code : -

Using all those five learning models which we discussed earlier for predicting.

**Using buit-in Iris dataset from Scikit-learn**

- This dataset contains 150 samples of iris flowers.

- Each sample has 4 features:
    - Sepal length (cm)
    - Sepal width (cm)
    - Petal length (cm)
    - Petal width (cm)

- Each sample belongs to one of 3 species:
    - Setosa (label = 0)
    - Versicolor (label = 1)
    - Virginica (label = 2)

In [1]:
from sklearn.datasets import load_iris
import pandas as pd

# Load dataset
iris = load_iris()

# Create DataFrame
df = pd.DataFrame(data=iris.data, columns=iris.feature_names)
df['target'] = iris.target

print(df.head()) # View dataset

   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0  


# Preprocess the data

In [2]:
# Split features and labels
X = df.drop('target', axis=1)  # Feature data
y = df['target']               # Labels

# Split into training and testing sets
from sklearn.model_selection import train_test_split

# Split the data (80% for training, 20% for testing)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=30)

# Code to Train and Predict

## 1. Linear Regression

In [3]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score, classification_report
import numpy as np

# Create the model
LR_model = LinearRegression()

# Train the model on training data
LR_model.fit(X_train, y_train)

# Predict labels for the test set
y_pred_lin = LR_model.predict(X_test)

# Since it's output is continuous numbers, we round them to nearest class
y_pred_round = np.round(y_pred_lin).astype(int)

# Clip the predictions to stay within valid label range [0, 2]
y_pred_clipped = np.clip(y_pred_round, 0, 2)

### Accuracy of the model

In [4]:
acc_lin = accuracy_score(y_test, y_pred_clipped)
print("Accuracy of Linear-Regression :", acc_lin)

# See a per-class breakdown
print(classification_report(y_test, y_pred_clipped, target_names=iris.target_names))

Accuracy of Linear-Regression : 0.9333333333333333
              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        12
  versicolor       0.89      0.89      0.89         9
   virginica       0.89      0.89      0.89         9

    accuracy                           0.93        30
   macro avg       0.93      0.93      0.93        30
weighted avg       0.93      0.93      0.93        30



### Evaluation

Accuracy: 93.3%
Predicted labels: Mostly correct, but not perfect
Evaluation: Misclassifies some versicolor/virginica samples.

It is not meant for classification but for continuous values. We are rounding these values to nearest classes.
That's why : 
- It is working for classes that are well seperated numerically (Setosa).
- Stuggles when class boundaries arent linear. (btw versicolor and virginica)

**Shouldn't be used for classification. Worked here only because the the probelm is simple and setosa is well seperated.**

## 2. Logistic Regression

In [5]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# Create the model
log_reg = LogisticRegression(
    max_iter=200,      # allow enough iterations to converge
)

# Train (fit) on the training data
log_reg.fit(X_train, y_train)

# Predict labels for the test set
y_pred_log = log_reg.predict(X_test)

In [6]:
# Evaluate
acc_log = accuracy_score(y_test, y_pred_log)
print("Accuracy of Logistic-Regression :", acc_log)

# See a per-class breakdown
print(classification_report(y_test, y_pred_log, target_names=iris.target_names))

Accuracy of Logistic-Regression : 0.9666666666666667
              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        12
  versicolor       1.00      0.89      0.94         9
   virginica       0.90      1.00      0.95         9

    accuracy                           0.97        30
   macro avg       0.97      0.96      0.96        30
weighted avg       0.97      0.97      0.97        30



### Evaluation

Accuracy: 96.7%
Slighly low recall for Virgenica

**Why this performance?**
Logistic regression draws linear boundaries between classes.

Works very well when classes are linearly separable — and Iris almost is (except versicolor vs virginica overlap).

What this tells about the data?
The dataset is mostly linearly separable, except a few cases between versicolor and virginica.

## 3. Decision Tree

In [7]:
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.metrics import accuracy_score, classification_report

# Create the model
dec_tree = DecisionTreeClassifier(
    criterion="gini",   # or "entropy"
    max_depth=3,              # prevent overfitting
    min_samples_split=4,      # don’t split unless ≥4 samples
    min_samples_leaf=2,       # each leaf should have ≥2 samples
    random_state=30     # reproducible splits
)

# Fit (train) on the training data
dec_tree.fit(X_train, y_train)

# Predict labels for the test set
y_pred_dt = dec_tree.predict(X_test)

In [8]:
# Evaluate
acc_dt = accuracy_score(y_test, y_pred_dt)
print("Accuracy of Decision-Tree:", acc_dt)
print(classification_report(y_test, y_pred_dt, target_names=iris.target_names))

Accuracy of Decision-Tree: 0.9333333333333333
              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        12
  versicolor       0.89      0.89      0.89         9
   virginica       0.89      0.89      0.89         9

    accuracy                           0.93        30
   macro avg       0.93      0.93      0.93        30
weighted avg       0.93      0.93      0.93        30



### Evaluation of this model

Accuracy: 93.3%

Why this performance?
Trees split based on thresholds of feature values (like petal length < x).
In Iris, a few splits are enough to classify most points correctly — but they may still make errors on overlapping samples.

**What it says about the data:**

Decision boundaries based on specific feature thresholds are enough to distinguish most classes.
But the noise/overlap between class 1 and 2 causes misclassifications.


## 4. Support Vector Machine

In [9]:
from sklearn.svm import SVC

# Create the model
svm = SVC(kernel='linear', C=1.0, random_state=30)

# Train the model
svm.fit(X_train, y_train)

# Predict on the test data
y_pred_svm = svm.predict(X_test)

In [10]:
# Evaluate
acc_svm = accuracy_score(y_test, y_pred_svm)
print("Accuracy of SVM :", acc_svm)
print(classification_report(y_test, y_pred_svm, target_names=iris.target_names))

Accuracy of SVM : 0.9666666666666667
              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        12
  versicolor       1.00      0.89      0.94         9
   virginica       0.90      1.00      0.95         9

    accuracy                           0.97        30
   macro avg       0.97      0.96      0.96        30
weighted avg       0.97      0.97      0.97        30



### Evaluation

Accuracy: 96.7% - same as logistic regression

Same interpretation.

## k-Nearest Neighbors

In [11]:
from sklearn.neighbors import KNeighborsClassifier

# Create the model
k_nn = KNeighborsClassifier(n_neighbors=5)  # 5 nearest neighbors

# Train the model (just memorizes training data)
k_nn.fit(X_train, y_train)

# Predict on test set
y_pred_k_nn = k_nn.predict(X_test)

In [12]:
# Evaluate
acc_k_nn = accuracy_score(y_test, y_pred_k_nn)
print("Accuracy of k-Nearest Neighbors :", acc_k_nn)
print(classification_report(y_test, y_pred_k_nn, target_names=iris.target_names))

Accuracy of k-Nearest Neighbors : 0.9333333333333333
              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        12
  versicolor       1.00      0.78      0.88         9
   virginica       0.82      1.00      0.90         9

    accuracy                           0.93        30
   macro avg       0.94      0.93      0.92        30
weighted avg       0.95      0.93      0.93        30



Accuracy: 93.3%

What this tells about the data:

Local neighborhood for Setosa is clean → perfect accuracy

But Versicolor vs Virginica overlap in feature space → confusion in neighborhood

# Regression Problem

Similar to the previous one, but here we are trying to solve a regression problem instead of classification.

We are using another built-in real world regression dataset - California_housing

In [13]:
# Combined code for training four algorithms
 
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import StandardScaler
import pandas as pd

# Load dataset
X, y = fetch_california_housing(return_X_y=True, as_frame=True)

# Optional: Standardize features (important for SVR, KNN, etc.)
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# Define models
models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree": DecisionTreeRegressor(random_state=30),
    "KNN Regressor": KNeighborsRegressor(n_neighbors=5),
    "SVR (RBF Kernel)": SVR(kernel='rbf')
}

# Train and evaluate
results = []

for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)

    results.append({
        "Model": name,
        "MSE": round(mse, 2),
        "R² Score": round(r2, 2)
    })

# Show results
df_results = pd.DataFrame(results)
print(df_results)

               Model   MSE  R² Score
0  Linear Regression  0.56      0.58
1      Decision Tree  0.50      0.62
2      KNN Regressor  0.43      0.67
3   SVR (RBF Kernel)  0.36      0.73


# Evaluation : -

1. SVR (RBF Kernel) – Best MSE = 0.36

- Most accurate model overall.

- This low MSE means the model is consistently close to the true values.

- The RBF kernel captures nonlinear relationships smoothly.

- Great when the data has complex, curved, or indirect patterns.


**The data likely has nonlinear dependencies.**

2. KNN Regressor – MSE = 0.43

- Second-best performance and very close to SVR.

- Predicts values based on averages of nearby training samples.

**The dataset has some local structure or clusters where nearby samples tend to have similar outputs.**

3. Decision Tree – MSE = 0.50

- Less precise than KNN or SVR.

- Predicts using stepwise thresholds, which might miss finer variations.

- Can create jumps in predictions instead of smooth transitions.

**The data has some natural thresholds or decision points. Not good but atleast better than Linear Regression**

4. Linear Regression – MSE = 0.56

- Worst performer here.

- Assumes a straight-line relationship between inputs and target.

- MSE is higher because it cannot handle curved or complex patterns.

- Still not terrible — means some linear trend is present, just not enough.


**The model is too simple for this dataset — linear assumptions don't capture the full pattern.**